cmd/erasure-healfile.go

```diff
@@ -36,48 +50,72 @@ func (s ErasureStorage) HealFile(offlineDisks []StorageAPI, volume, path string,
 			f.Checksums[i] = checksums[i]
 		}
 	}
-	blocks := make([][]byte, len(s.disks))
+
+	// Scan part files in disk, block-by-block reconstruct it and
+	// write to stale disks.
 	chunksize := getChunkSize(blocksize, s.dataBlocks)
-	for offset := int64(0); offset < size; offset += blocksize {
-		if size < blocksize {
-			blocksize = size
+	var chunkOffset, blockOffset int64
+	for blockOffset = 0; blockOffset < size; blockOffset += blocksize {
+		// last iteration may have less than blocksize data
+		// left, so chunksize needs to be recomputed.
+		if size < blockOffset+blocksize {
+			blocksize = size - blockOffset
 			chunksize = getChunkSize(blocksize, s.dataBlocks)
 		}
+
+		// read a chunk from each disk, until we have
+		// `s.dataBlocks` number of chunks set to non-nil in
+		// `blocks`
+		blocks := make([][]byte, len(s.disks))
 		numReads := 0
 		for i, disk := range s.disks {
-			if disk != OfflineDisk {
-				if blocks[i] == nil {
-					blocks[i] = make([]byte, chunksize)
-				}
-				blocks[i] = blocks[i][:chunksize]
-				if !verifiers[i].IsVerified() {
-					_, err = disk.ReadFileWithVerify(volume, path, offset, blocks[i], verifiers[i])
-				} else {
-					_, err = disk.ReadFile(volume, path, offset, blocks[i])
-				}
-				if err != nil {
-					blocks[i] = nil
-				} else {
-					numReads++
-				}
-				if numReads == s.dataBlocks { // we have enough data to reconstruct
-					break
-				}
+			if disk == OfflineDisk {
+				continue
+			}
+			blocks[i] = make([]byte, chunksize)
+			if !verifiers[i].IsVerified() {
+				_, err = disk.ReadFileWithVerify(volume, path,
+					chunkOffset, blocks[i], verifiers[i])
+			} else {
+				_, err = disk.ReadFile(volume, path, chunkOffset,
+					blocks[i])
+			}
+			if err != nil {
+				blocks[i] = nil
+				continue
+			}
+
+			numReads++
+			if numReads == s.dataBlocks {
+				// we have enough data to reconstruct
+				break
 			}
 		}
+
+		// reconstruct data - this computes all data and parity shards
 		if err = s.ErasureDecodeDataAndParityBlocks(blocks); err != nil {
 			return f, err
 		}
+
+		// write computed shards as chunks on file in each
+		// stale disk
 		for i, disk := range s.disks {
 			if disk != OfflineDisk {
 				continue
 			}
-			if err = offlineDisks[i].AppendFile(healVolume, healPath, blocks[i]); err != nil {
+
+			err = staleDisks[i].AppendFile(healVolume, healPath,
+				blocks[i])
+			if err != nil {
 				return f, traceError(err)
 			}
 			hashers[i].Write(blocks[i])
 		}
+
+		chunkOffset += chunksize
```

Comment:
Can we move this up? It is highly related to blockOffset but is incremented on the "other end" of the function....

Reply:
Sure, I'll move this up as far as possible.

It is also possible to write like `for i, j = 0, 0; i<10; i, j = i+di, j+dj {` but this way does not look good with longer variable names.